### ISSUE TITLE
Feature importance / model inspection

### ISSUE THREAD

**Issue by baggepinnen** [alan-turing-institute/MLJ.jl#403]
It would be nice to have some integrated tools for model inspection and feature importance. Below are some links to resources and what's available in scikit learn.

Scikit learn exposes a number of tools for understanding the relative importance of features in a dataset. These tools are general in the sense that they can be made to work with many different kinds of models. They are organized in a module called "Inspection" which I find fitting, since they all allow the user to somehow understand or *inspect* the result of fitting a model in other ways than simply measuring the error/accuracy. Some of them are linked below

- [Permutation feature importance in scikit](https://scikit-learn.org/stable/modules/permutation_importance.html). This one is very general.
- [random forest FI in scikit](https://scikit-learn.org/stable/auto_examples/ensemble/plot_forest_importances.html)
- [partial dependence plot](https://scikit-learn.org/stable/modules/partial_dependence.html) this is not really a feature-importance tool, but nevertheless an interesting tool to have to understand the result of the fitted model.
- [Drop Column feature importance](https://towardsdatascience.com/explaining-feature-importance-by-example-of-a-random-forest-d9166011959e) This one is simple, train a model without a feature and see how much worse it gets.
- [LIME](https://towardsdatascience.com/understanding-model-predictions-with-lime-a582fdff3a3b) A general and popular method model-agnostic tool for understanding the impact of perturbing inputs.
- [A Unified Approach to Interpreting Model Predictions](http://papers.nips.cc/paper/7062-a-unified-approach-to-interpreting-model-predictions.pdf). A recent paper on model inspection, going through LIME and other related methods.

**Comment by tlienart** [comment]
Thanks for this.

For bagging ensembles it's reasonably straightforward. Some models we interface with  also have it (e.g. XGBoost) and so it can  just be part of  the interface (via `report`), I'll have a look into this for xgb. 

Support for permutation/drop FI seems reasonably easy, there's just a question as to _where_ the implementation would go, maybe a comparable "model(s) inspection" module or package in MLJ or something of the sorts.

The rest of your suggestions are a bit trickier. LIME is nice but basically is an entire package; like [shap](https://github.com/slundberg/shap) which is quoted in the article at your last point.

**Comment by ablaom** [comment]
Feature importance is an interesting one because most of the measures out there are rather ad-hoc and model dependent. That is, the very *definition* of feature importance depends on the model (eg, absolute value of a coefficient in a linear model makes no sense for a decision tree). And for certain models, eg trees and random forests, there are several inequivalent methods in common use. The paper cited above on shap describes an approach that is really model independent; unless someone is aware of another such approach, I suggest any generic MLJ tool follow that approach. There is already some implementation of SHAP in python, if I remember correctly.

**Comment by tlienart** [comment]
The recently created https://github.com/nredell/ShapML.jl may also be a very nice add (already compatible with MLJ as far as I can see) cc @nredell

**Comment by nredell** [comment]
My plans for ShapML can be found on the Discourse site--[https://discourse.julialang.org/t/ml-feature-importance-in-julia/17196/12](https://discourse.julialang.org/t/ml-feature-importance-in-julia/17196/12)--, but I'm posting here for posterity sake.

Just sitting down for the first refactor/feature additions today. I'll code with these guidelines in mind ([https://github.com/invenia/BlueStyle](https://github.com/invenia/BlueStyle)) as well as take a trip through the MLJ code base. And if a general feature importance package pops up in the future, I wouldn't be opposed to helping fold ShapML in if it's up to par and hasn't expanded too much by then.

**Comment by ablaom** [comment]
cc @sjvollmer (for summer FAIRness student, if not already aware)

**Comment by ablaom** [comment]
My current inclination is to see if this can be satisfactorily addressed with third party packages, such as the Shapley one. A POC would make a great [MLJTutorial](https://github.com/alan-turing-institute/MLJTutorials).

If something more integrated makes sense, though, l'm interested to here about it.

**Comment by vishalhedgevantage** [comment]
Any update on feature importance integration

**Comment by ablaom** [comment]
@vishalhedgevantage There are some GSoC students working on better integration of interpretable machine learning (LIME/Shapley). And there is [this issue](https://github.com/alan-turing-institute/MLJ.jl/issues/747), which I opened to support [recursive feature elimination](https://github.com/alan-turing-institute/MLJ.jl/issues/747). However, the volunteer who had expressed an interest in the latter must of got busy with other things...

**Comment by Moelf** [comment]
>For bagging ensembles it's reasonably straightforward. Some models we interface with also have it (e.g. XGBoost) and so it can just be part of the interface (via report), I'll have a look into this for xgb.

Any movement on this?

**Comment by ablaom** [comment]
EvoTrees.jl is a pure julia gradient tree boosting algorithm which already has a lot of functionality to be found in XGBoost.jl and, as far as I can tell, is implementing basically the same algorithm. It does not have all the bells and whistles, but it is being actively developed.
